### ISSUE TITLE
Jython deadlock

### ISSUE THREAD

**Issue by kasemir** [ControlSystemStudio/phoebus#1310]
Observed before the jython 2.7.2 update, i.e. still with 2.7.1, after opening several displays that all include rules and scripts.

```
Found one Java-level deadlock:
=============================
"ScriptSupport":
  waiting for ownable synchronizer 0x0000000702275c60, (a java.util.concurrent.locks.ReentrantLock$NonfairSync),
  which is held by "ScriptSupport"

"ScriptSupport":
  waiting to lock monitor 0x00000000079f0d00 (object 0x000000070215e8b0, a org.python.core.PySystemState),
  which is held by "ScriptSupport"

"ScriptSupport":
  waiting for ownable synchronizer 0x0000000702275c60, (a java.util.concurrent.locks.ReentrantLock$NonfairSync),
  which is held by "ScriptSupport"

Java stack information for the threads listed above:
===================================================
"ScriptSupport":
	at jdk.internal.misc.Unsafe.park(java.base@14/Native Method)
	- parking to wait for  <0x0000000702275c60> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(java.base@14/LockSupport.java:211)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@14/AbstractQueuedSynchronizer.java:714)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@14/AbstractQueuedSynchronizer.java:937)
	at java.util.concurrent.locks.ReentrantLock$Sync.lock(java.base@14/ReentrantLock.java:153)
	at java.util.concurrent.locks.ReentrantLock.lock(java.base@14/ReentrantLock.java:322)
	at org.python.core.imp.importName(imp.java:1055)
	at org.python.core.ImportFunction.__call__(__builtin__.java:1280)
	at org.python.core.PyObject.__call__(PyObject.java:450)
	at org.python.core.__builtin__.__import__(__builtin__.java:1232)
	at org.python.core.imp.importFromAs(imp.java:1149)
	at org.python.core.imp.importFrom(imp.java:1124)
	at org.python.pycode._pyx22.f$0(label:Label_19:Visible.rule.py:19)
	at org.python.pycode._pyx22.call_function(label:Label_19:Visible.rule.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.Py.runCode(Py.java:1614)
	at org.python.core.Py.exec(Py.java:1658)
	at org.python.util.PythonInterpreter.exec(PythonInterpreter.java:276)
	at org.csstudio.display.builder.runtime.script.internal.JythonScriptSupport.lambda$submit$0(JythonScriptSupport.java:234)
	at org.csstudio.display.builder.runtime.script.internal.JythonScriptSupport$$Lambda$1550/0x00000007c09d7c40.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(java.base@14/FutureTask.java:264)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@14/ThreadPoolExecutor.java:1130)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@14/ThreadPoolExecutor.java:630)
	at java.lang.Thread.run(java.base@14/Thread.java:832)
"ScriptSupport":
	at org.python.core.PySystemState.getCodecState(PySystemState.java:357)
	- waiting to lock <0x000000070215e8b0> (a org.python.core.PySystemState)
	at org.python.core.codecs.getDefaultEncoding(codecs.java:38)
	at org.python.core.codecs.decode(codecs.java:83)
	at org.python.core.PyUnicode.unicode_new(PyUnicode.java:612)
	at org.python.core.PyUnicode$exposed___new__.new_impl(Unknown Source)
	at org.python.core.PyType.invokeNew(PyType.java:583)
	at org.python.core.PyType.type___call__(PyType.java:1815)
	at org.python.core.PyType.__call__(PyType.java:1805)
	at org.python.core.PyObject.__call__(PyObject.java:480)
	at org.python.core.PyObject.__call__(PyObject.java:484)
	at sre_compile$py.f$0(/home/controls/css/product-sns-2020-05-06/lib/jython-standalone-2.7.1.jar/Lib/sre_compile.py:506)
	at sre_compile$py.call_function(/home/controls/css/product-sns-2020-05-06/lib/jython-standalone-2.7.1.jar/Lib/sre_compile.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.imp.createFromCode(imp.java:436)
	at org.python.core.util.importer.importer_load_module(importer.java:116)
	at org.python.modules.zipimport.zipimporter.zipimporter_load_module(zipimporter.java:170)
	at org.python.modules.zipimport.zipimporter$zipimporter_load_module_exposer.__call__(Unknown Source)
	at org.python.core.PyBuiltinMethodNarrow.__call__(PyBuiltinMethodNarrow.java:46)
	at org.python.core.imp.loadFromLoader(imp.java:593)
	at org.python.core.imp.find_module(imp.java:547)
	at org.python.core.imp.import_next(imp.java:838)
	at org.python.core.imp.import_module_level(imp.java:957)
	at org.python.core.imp.importName(imp.java:1057)
	at org.python.core.ImportFunction.__call__(__builtin__.java:1280)
	at org.python.core.PyObject.__call__(PyObject.java:450)
	at org.python.core.__builtin__.__import__(__builtin__.java:1232)
	at org.python.core.imp.importOne(imp.java:1076)
	at re$py.f$0(/home/controls/css/product-sns-2020-05-06/lib/jython-standalone-2.7.1.jar/Lib/re.py:290)
	at re$py.call_function(/home/controls/css/product-sns-2020-05-06/lib/jython-standalone-2.7.1.jar/Lib/re.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.imp.createFromCode(imp.java:436)
	at org.python.core.util.importer.importer_load_module(importer.java:116)
	at org.python.modules.zipimport.zipimporter.zipimporter_load_module(zipimporter.java:170)
	at org.python.modules.zipimport.zipimporter$zipimporter_load_module_exposer.__call__(Unknown Source)
	at org.python.core.PyBuiltinMethodNarrow.__call__(PyBuiltinMethodNarrow.java:46)
	at org.python.core.imp.loadFromLoader(imp.java:593)
	at org.python.core.imp.find_module(imp.java:547)
	at org.python.core.imp.import_next(imp.java:838)
	at org.python.core.imp.import_first(imp.java:872)
	at org.python.core.imp.import_module_level(imp.java:964)
	at org.python.core.imp.importName(imp.java:1057)
	at org.python.core.ImportFunction.__call__(__builtin__.java:1280)
	at org.python.core.PyObject.__call__(PyObject.java:450)
	at org.python.core.__builtin__.__import__(__builtin__.java:1232)
	at org.python.core.imp.importOne(imp.java:1076)
	at xml.etree.ElementTree$py.f$0(/home/controls/css/product-sns-2020-05-06/lib/jython-standalone-2.7.1.jar/Lib/xml/etree/ElementTree.py:1680)
	at xml.etree.ElementTree$py.call_function(/home/controls/css/product-sns-2020-05-06/lib/jython-standalone-2.7.1.jar/Lib/xml/etree/ElementTree.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.imp.createFromCode(imp.java:436)
	at org.python.core.util.importer.importer_load_module(importer.java:116)
	at org.python.modules.zipimport.zipimporter.zipimporter_load_module(zipimporter.java:170)
	at org.python.modules.zipimport.zipimporter$zipimporter_load_module_exposer.__call__(Unknown Source)
	at org.python.core.PyBuiltinMethodNarrow.__call__(PyBuiltinMethodNarrow.java:46)
	at org.python.core.imp.loadFromLoader(imp.java:593)
	at org.python.core.imp.find_module(imp.java:547)
	at org.python.core.PyModule.impAttr(PyModule.java:111)
	at org.python.core.imp.import_next(imp.java:840)
	at org.python.core.imp.import_logic(imp.java:905)
	at org.python.core.imp.import_module_level(imp.java:970)
	at org.python.core.imp.importName(imp.java:1057)
	at org.python.core.ImportFunction.__call__(__builtin__.java:1280)
	at org.python.core.PyObject.__call__(PyObject.java:450)
	at org.python.core.__builtin__.__import__(__builtin__.java:1232)
	at org.python.core.imp.importFromAs(imp.java:1149)
	at org.python.core.imp.importFrom(imp.java:1124)
	at xml.etree.cElementTree$py.f$0(/home/controls/css/product-sns-2020-05-06/lib/jython-standalone-2.7.1.jar/Lib/xml/etree/cElementTree.py:3)
	at xml.etree.cElementTree$py.call_function(/home/controls/css/product-sns-2020-05-06/lib/jython-standalone-2.7.1.jar/Lib/xml/etree/cElementTree.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.imp.createFromCode(imp.java:436)
	at org.python.core.util.importer.importer_load_module(importer.java:116)
	at org.python.modules.zipimport.zipimporter.zipimporter_load_module(zipimporter.java:170)
	at org.python.modules.zipimport.zipimporter$zipimporter_load_module_exposer.__call__(Unknown Source)
	at org.python.core.PyBuiltinMethodNarrow.__call__(PyBuiltinMethodNarrow.java:46)
	at org.python.core.imp.loadFromLoader(imp.java:593)
	at org.python.core.imp.find_module(imp.java:547)
	at org.python.core.PyModule.impAttr(PyModule.java:111)
	at org.python.core.imp.import_next(imp.java:840)
	at org.python.core.imp.import_logic(imp.java:905)
	at org.python.core.imp.import_module_level(imp.java:970)
	at org.python.core.imp.importName(imp.java:1057)
	at org.python.core.ImportFunction.__call__(__builtin__.java:1280)
	at org.python.core.PyObject.__call__(PyObject.java:450)
	at org.python.core.__builtin__.__import__(__builtin__.java:1232)
	at org.python.core.imp.importOneAs(imp.java:1093)
	at scan.client.scanclient$py.f$0(/home/controls/css/PyScanClient/build/lib/scan/client/scanclient.py:125)
	at scan.client.scanclient$py.call_function(/home/controls/css/PyScanClient/build/lib/scan/client/scanclient.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.imp.createFromCode(imp.java:436)
	at org.python.core.imp.createFromPyClass(imp.java:237)
	at org.python.core.imp.createFromPyClass(imp.java:205)
	at org.python.core.imp.loadFromSource(imp.java:657)
	at org.python.core.imp.find_module(imp.java:551)
	at org.python.core.PyModule.impAttr(PyModule.java:111)
	at org.python.core.imp.import_next(imp.java:840)
	at org.python.core.imp.import_module_level(imp.java:957)
	at org.python.core.imp.importName(imp.java:1057)
	at org.python.core.ImportFunction.__call__(__builtin__.java:1280)
	at org.python.core.PyObject.__call__(PyObject.java:450)
	at org.python.core.__builtin__.__import__(__builtin__.java:1232)
	at org.python.core.imp.importFromAs(imp.java:1149)
	at org.python.core.imp.importFrom(imp.java:1124)
	at scan.client$py.f$0(/home/controls/css/PyScanClient/build/lib/scan/client/__init__.py:3)
	at scan.client$py.call_function(/home/controls/css/PyScanClient/build/lib/scan/client/__init__.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.imp.createFromCode(imp.java:436)
	at org.python.core.imp.createFromPyClass(imp.java:237)
	at org.python.core.imp.createFromPyClass(imp.java:205)
	at org.python.core.imp.loadFromSource(imp.java:657)
	at org.python.core.imp.find_module(imp.java:551)
	at org.python.core.PyModule.impAttr(PyModule.java:111)
	at org.python.core.imp.import_next(imp.java:840)
	at org.python.core.imp.import_logic(imp.java:905)
	at org.python.core.imp.import_module_level(imp.java:970)
	at org.python.core.imp.importName(imp.java:1057)
	at org.python.core.ImportFunction.__call__(__builtin__.java:1280)
	at org.python.core.PyObject.__call__(PyObject.java:450)
	at org.python.core.__builtin__.__import__(__builtin__.java:1232)
	at org.python.core.imp.importAll(imp.java:1176)
	at scan$py.f$0(/home/controls/css/PyScanClient/build/lib/scan/__init__.py:6)
	at scan$py.call_function(/home/controls/css/PyScanClient/build/lib/scan/__init__.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.imp.createFromCode(imp.java:436)
	at org.python.core.imp.createFromPyClass(imp.java:237)
	at org.python.core.imp.createFromPyClass(imp.java:205)
	at org.python.core.imp.loadFromSource(imp.java:657)
	at org.python.core.imp.find_module(imp.java:551)
	at org.python.core.imp.import_next(imp.java:838)
	at org.python.core.imp.import_module_level(imp.java:957)
	at org.python.core.imp.importName(imp.java:1057)
	at org.python.core.ImportFunction.__call__(__builtin__.java:1280)
	at org.python.core.PyObject.__call__(PyObject.java:450)
	at org.python.core.__builtin__.__import__(__builtin__.java:1232)
	at org.python.core.imp.importFromAs(imp.java:1149)
	at org.python.core.imp.importFrom(imp.java:1124)
	at org.python.pycode._pyx29.f$0(scripts_bob/tablescan_init.py:55)
	at org.python.pycode._pyx29.call_function(scripts_bob/tablescan_init.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.Py.runCode(Py.java:1614)
	at org.python.core.Py.exec(Py.java:1658)
	at org.python.util.PythonInterpreter.exec(PythonInterpreter.java:276)
	at org.csstudio.display.builder.runtime.script.internal.JythonScriptSupport.lambda$submit$0(JythonScriptSupport.java:234)
	at org.csstudio.display.builder.runtime.script.internal.JythonScriptSupport$$Lambda$1550/0x00000007c09d7c40.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(java.base@14/FutureTask.java:264)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@14/ThreadPoolExecutor.java:1130)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@14/ThreadPoolExecutor.java:630)
	at java.lang.Thread.run(java.base@14/Thread.java:832)
"ScriptSupport":
	at jdk.internal.misc.Unsafe.park(java.base@14/Native Method)
	- parking to wait for  <0x0000000702275c60> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(java.base@14/LockSupport.java:211)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@14/AbstractQueuedSynchronizer.java:714)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@14/AbstractQueuedSynchronizer.java:937)
	at java.util.concurrent.locks.ReentrantLock$Sync.lock(java.base@14/ReentrantLock.java:153)
	at java.util.concurrent.locks.ReentrantLock.lock(java.base@14/ReentrantLock.java:322)
	at org.python.core.imp.load(imp.java:719)
	at org.python.core.PySystemState.getCodecState(PySystemState.java:360)
	- locked <0x000000070215e8b0> (a org.python.core.PySystemState)
	at org.python.core.codecs.getDefaultEncoding(codecs.java:38)
	at org.python.core.codecs.decode(codecs.java:83)
	at org.python.core.PyString.decode(PyString.java:4001)
	at org.python.core.PyString.decode(PyString.java:3993)
	at org.python.core.PyString.str___add__(PyString.java:1025)
	at org.python.core.PyString.__add__(PyString.java:1013)
	at org.python.core.PyObject._basic_add(PyObject.java:2151)
	at org.python.core.PyObject._add(PyObject.java:2137)
	at org.python.pycode._pyx30.f$0(scripts/ScriptScan_load.py:30)
	at org.python.pycode._pyx30.call_function(scripts/ScriptScan_load.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:171)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.Py.runCode(Py.java:1614)
	at org.python.core.Py.exec(Py.java:1658)
	at org.python.util.PythonInterpreter.exec(PythonInterpreter.java:276)
	at org.csstudio.display.builder.runtime.script.internal.JythonScriptSupport.lambda$submit$0(JythonScriptSupport.java:234)
	at org.csstudio.display.builder.runtime.script.internal.JythonScriptSupport$$Lambda$1550/0x00000007c09d7c40.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(java.base@14/FutureTask.java:264)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@14/ThreadPoolExecutor.java:1130)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@14/ThreadPoolExecutor.java:630)
	at java.lang.Thread.run(java.base@14/Thread.java:832)

Found 1 deadlock.
```

**Comment by kasemir** [comment]
Basic deadlock is between the 'import lock',
https://jython.readthedocs.io/en/latest/Concurrency/#module-import-lock,
and a synchronized section method in `PySystemState`.


 - 3rd `ScriptSupport` thread executes string addition, needs encoding info, PySystemState.getCodecState locks PySystemState, and then runs "import encoding", which takes the import lock.
--> Locks PySystemState, then ImportLock

 - 2nd `ScriptSupport` thread executes an initial import, i.e. takes import lock. That import results in calling `__init__.py` code, which imports other code, creates python types, and eventually needs to also call PySystemState.getCodecState 
--> Locks ImportLock, then PySystemState


Taking the same two locks in different order causes the deadlock.

With jython 2.7.2, PySystemState.getCodecState looks like this, it will only once result in taking the Import Lock:
```
public synchronized codecs.CodecState getCodecState() {
    if (codecState == null) {
        codecState = new codecs.CodecState();
        try {
            imp.load("encodings");
        } catch (PyException exc) {
            if (exc.type != Py.ImportError) {
                throw exc;
            }
        }
    }
    return codecState;
}
```

With the latest display builder code, a breakpoint on the `imp.load("encodings")` line shows that it is only called once, from the initial `JythonScriptSupport.compile()` call when adding to the python path. It is thus called before executing any Jython scripts.
During later execution of scripts, `getCodecState` always uses the already initialized state, not performing any more `imp.load` as previously seen in the 3rd `ScriptSupport` thread.
This exact deadlock should thus no longer occur after the 2.7.2 update.

**Comment by kasemir** [comment]
Tested recently added `issues/LongScript*.bob` examples with previous jython version 2.7.1: PySystemState.getCodecState is not called during initial JythonScriptSupport.compile() but later during execution of script, as in the observed deadlock. The jython update thus avoids this deadlock.

**Comment by kasemir** [comment]
Fixed by #1306
